### ISSUE TITLE
Expose NSS VerifyCodeSigningCertificateChain in `rc_crypto` (requires NSS 3.66)

### ISSUE THREAD

**Issue by data-sync-user** [mozilla/uniffi-rs#886]
Periodic update.  As part of the week, need to review the existing documentation.

The rc_crypto crate is probably our scariest crate, both because it's crypto-related and because it's got some weird licensing stuff going on. It's been a while since any of us had to deal with said scariness, so it's a good time to revisit that crate with fresh eyes and see if there are any docs that need to be updated (or brought into existence).

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

leplatrem is this the right issue to track rc_crypto getting the new nss api that was added here ( https://bugzilla.mozilla.org/show_bug.cgi?id=1709291 )?

**Comment by data-sync-user** [comment]
➤ Mathieu Leplatre commented:

> leplatrem is this the right issue to track rc_crypto getting the new nss api that was added here?

Yes!

Ryan updated the documentation about the process of upgrading NSS, and I still have pending work to do before I can give any ETA :pray:

**Comment by data-sync-user** [comment]
➤ Ryan Kelly commented:

The NSS version update was done in https://github.com/mozilla/application-services/pull/4125. I believe we still have work to do to expose the new function in the rc_crypto API surface. I wasn't able to find a separate issue tracking that work - leplatrem do we have one, or should we consider changing the title of this issue to cover that work?

Anyway, some notes on what remains to be done before we can use the new function from Rust consumers:

  * We first need to expose the new function from the `nss_sys` crate ( https://github.com/mozilla/application-services/tree/main/components/support/rc_crypto/nss/nss_sys ). This crate more-or-less directly exposes the underlying NSS APIs into Rust, so it's where we need to start.
  * Now you might be looking at the files in this crate and thinking "this looks like it was autogenerated from the NSS header files, but how do I regenerate it?". That's a very fair question to which I was trying to find the answer earlier today. It seems that the scripts for generating these files were removed in https://github.com/mozilla/application-services/pull/2974 when we decided to check in the resulting generated bindings. I'll take a look at restoring them somewhere so it's easier to make this kind of change in the future :-/

  * Next, we need to wrap the new function into a safe Rust API and expose it from this `nss` crate ( https://github.com/mozilla/application-services/tree/main/components/support/rc_crypto/nss ).
  * I don't have any strongly-held opinions on how it should be named. I see that the new function comes in a new NSS header file named pkixc, so maybe in a new submodule like nss::pkixc::verify_code_signing_certificate_chain? I haven't looked at the new function in enough detail to understand how it should fit in here.

  * Finally, we would want to expose a yet-higher-level Rust API through the `rc_crypto` crate ( https://github.com/mozilla/application-services/tree/main/components/support/rc_crypto ).
  * Must of the rc_crypto API is based on Rust's ring crate, but I don't think it has anything specifically for this. I'm thinking perhaps an rc_crypto::contentsignature submodule that mirrors some subset of this go module API ( https://pkg.go.dev/github.com/mozilla-services/autograph/verifier/contentsignature#Verify )?

**Comment by data-sync-user** [comment]
➤ Ryan Kelly commented:

> Now you might be looking at the files in this crate and thinking "this looks like it was autogenerated
> from the NSS header files, but how do I regenerate it?".

Turns out these are now artisanally hand-crafted bindings, but I've added some notes on how to edit them in https://github.com/mozilla/application-services/issues/4175
